### PR TITLE
add Engine.ResetRoutes() for clean existed routes

### PR DESCRIPTION
**Pull request by tsaikd**
Signed-off-by: tsaikd tsaikd@gmail.com

### CODE REVIEW

**Comment by thinkerou** [comment]
@tsaikd please re-commit the pull request to `master` branch, thanks!

**Comment by thinkerou** [comment]
NO, please squash and recommit master branch not develop branch.

File: gin.go [hunk 1]
```diff
@@ -281,6 +281,11 @@ func iterate(path, method string, routes RoutesInfo, root *node) RoutesInfo {
 	return routes
 }
 
+// ResetRoutes reset routes in Engine, used when application need to change routes dynamicly
```
Comment (dmarkham) [inline gin.go]: dynamically

File: gin.go [hunk 2]
```diff
@@ -281,6 +281,11 @@ func iterate(path, method string, routes RoutesInfo, root *node) RoutesInfo {
...
+func (engine *Engine) ResetRoutes() {
+	engine.trees = make(methodTrees, 0, 9)
+}
+
```
Comment (dmarkham) [inline gin.go]: If the point of this is to change the router while it's running. Is this thread safe?

**Comment by dmarkham** [comment]
`Engine.ResetRoutes()` Do we plan on this method being thread safe? This test fails the race detector currently. I did not test to see what it will do if you are hitting routes at a high rates and change it. 
 
If we choose to make it thread safe then I think it will snowball into `addRoute` also needing to be safe. I think the reason we are `ResetRoutes()` is to change the routes, and if the system is up and running while we are changing the routes what do we expect to happen?

```
func TestResetRoutes(t *testing.T) {
	router := New()
	router.addRoute("GET", "/", HandlersChain{func(_ *Context) {}})

	assert.Len(t, router.trees, 1)
	assert.NotNil(t, router.trees.get("GET"))
	assert.Nil(t, router.trees.get("POST"))
	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		router.ResetRoutes()
		wg.Done()
	}()
	wg.Add(1)
	go func() {
		router.ResetRoutes()
		wg.Done()
	}()
	wg.Wait()
	assert.Empty(t, router.trees)
	assert.Nil(t, router.trees.get("GET"))
	assert.Nil(t, router.trees.get("POST"))
}
```

**Comment by tsaikd** [comment]
I use it for unit test only. If the behavior or use case is not what you expect, it's ok for me to close this PR.

**Comment by dmarkham** [comment]
> I use it for unit test only. If the behavior or use case is not what you expect, it's ok for me to close this PR.

@tsaikd  Thank you for your use case! Wonder if there is a way can get this functionality for testing without making it a Public method?... What is the advantage of `ResetRoutes()` over just a new Router `router := gin.New()` in test cases?

**Comment by tsaikd** [comment]
I forgot the detail about this. It's a few years ago PR. Seems there were no `gin.New()` (only global routing)?

**Pull request closed without merging**